src/doc.cpp

```diff
@@ -1158,7 +1158,23 @@ void Doc::ConvertAnalyticalMarkupDoc(bool permanent)
 void Doc::TransposeDoc()
 {
     Transposer transposer;
-    transposer.SetTransposition(this->m_options->m_transpose.GetValue());
+    bool setTransp = transposer.SetTransposition(this->m_options->m_transpose.GetValue());
+    if (!setTransp) {
+        // Set transposition by key tonic.
+        // Detect the current key from the keysignature.
+        KeySig *keysig = dynamic_cast<KeySig *>(this->FindDescendantByType(KEYSIG, 9));
```

Comment:
This is the right approach, but you need to look for it in the Doc::m_scoreDef. That is, doing `this->m_scoreDef.FindDescendantByType(KEYSIG, 1)`. (I would look only in the first level, because we want to find the scoreDef@key.sig. (If there are different on staffDefs, this is a different issue, I guess)

Reply:
I changed it to `this->m_scoreDef.FindDescendantByType(KEYSIG, 3)`, as the musicXML import places the keySig inside a StaffDef. (I'm testing with [13b-KeySignatures-ChurchModes.xml](https://github.com/cuthbertLab/musicxmlTestSuite/blob/master/xmlFiles/13b-KeySignatures-ChurchModes.xml)).

```xml
<music>
<body>
  <mdiv xml:id="mdiv-0000002032342191">
    <score xml:id="score-0000001847798602">
      <scoreDef xml:id="scoredef-0000000735621517">
        <staffGrp xml:id="staffgrp-0000000088293271">
          <staffDef xml:id="staffdef-0000000842034478" n="1" lines="5" ppq="1" clef.shape="G" clef.line="2" meter.count="1" meter.unit="4">
            <keySig xml:id="keysig-0000000350207968" mode="major" sig="2s" />
```

If we guaranteed the keySig (or key.sig) was in the scoreDef, we could do

```c++
KeySig *keysig = this->m_scoreDef->GetKeySig();
```